frontend/src/components/dashboard/hooks/useManualGrouping.ts

```diff
@@ -0,0 +1,202 @@
+    );
+
+    idx = Math.floor(idx / numberOfCharacters);
+  } while (idx > 0);
+
+  return name.reverse().join("");
+};
+
+const isWithinGroup = (
+  group: Omit<ManualGroup, "label">,
+  x: number,
+  y: number,
+): boolean =>
+  group.minX <= x && x < group.maxX && group.minY <= y && y < group.maxY;
+
+interface IntermediateAnalysis extends DataPoint {
+  solution: CurrentAnalysis;
+  category: Category;
+  groupKey: string;
+}
+
+export const useManualGrouping = (
+  isAutomaticGrouping: boolean,
+  solutions: CurrentAnalysis[] | undefined,
+  filters: FilterCriterion[],
+  splits: ChartSplit[],
+  xAxis: AxesCriterionType,
+  yAxis: AxesCriterionType,
+): {
+  dataPoints: CategorizedDataPoints[];
+  groupAssignment: SolutionGroupAssignment[];
+  groups: ManualGroup[];
+} =>
+  useMemo(() => {
+    if (!solutions || isAutomaticGrouping) {
+      return {
+        dataPoints: [],
+        groupAssignment: [],
+        groups: [],
+      };
+    }
+
+    const groups: (Omit<ManualGroup, "label"> & { label?: string })[] = [];
+
+    const verticalSplits = splits
+      .filter((s) => s.type === SplitType.vertical)
+      .sort((a, b) => a.x - b.x);
+    const horizontalSplits = splits
+      .filter((s) => s.type === SplitType.horizontal)
+      .sort((a, b) => a.y - b.y);
+
+    let lastX = Number.NEGATIVE_INFINITY;
+
+    for (const vSplit of [
+      ...verticalSplits,
+      // add an artificial split to include the groups after the last vertical split
+      { type: SplitType.vertical, x: Number.POSITIVE_INFINITY },
+    ]) {
+      let lastY = Number.NEGATIVE_INFINITY;
+      for (const hSplit of [
+        ...horizontalSplits,
+        // add an artificial split to include the groups after the last horizontal split
+        {
+          type: SplitType.horizontal,
+          y: Number.POSITIVE_INFINITY,
+        },
+      ]) {
+        groups.push({
+          key: groups.length.toString(),
+          minX: lastX,
+          maxX: vSplit.x,
+          minY: lastY,
+          maxY: hSplit.y,
+        });
+
+        lastY = hSplit.y;
+      }
+
+      lastX = vSplit.x;
+    }
+
+    let usedGroupIdx = 0;
+
+    const solutionsByCategory = solutions
+      .map<IntermediateAnalysis>((solution) => {
+        const xAxisValue = getAxisAnalysisValue(xAxis, solution);
+        const yAxisValue = getAxisAnalysisValue(yAxis, solution);
+
+        const matchesAllFilters = filters
+          .map((f) => matchesFilter(f, solution))
+          .reduce(
+            (matchesAllFilters, matchesFilter) =>
+              matchesAllFilters && matchesFilter,
+            true,
+          );
+
+        const category = getCategory(solution, matchesAllFilters);
+
+        const group = groups.find((g) =>
+          isWithinGroup(g, xAxisValue, yAxisValue),
+        );
+
+        if (!group) {
+          throw new Error(
+            "All data points must be within a group - this error should never be thrown",
```

Comment:
Consider adding information about the solution here, so that we can later debug it, otherwise the message won't help fix production issues.